src/java/com/rapleaf/jack/queries/GenericQueryBuilder.java

```diff
@@ -0,0 +1,176 @@
+    return this;
+  }
+
+  public GenericQueryBuilder orderBy(Column column) {
+    genericQuery.addOrderCondition(new OrderCriterion(column, QueryOrder.ASC));
+    return this;
+  }
+
+  public GenericQueryBuilder limit(int offset, int limit) {
+    genericQuery.addLimitCondition(new LimitCriterion(offset, limit));
+    return this;
+  }
+
+  public GenericQueryBuilder limit(int limit) {
+    genericQuery.addLimitCondition(new LimitCriterion(limit));
+    return this;
+  }
+
+  public GenericQueryBuilder groupBy(Column column, Column... columns) {
+    genericQuery.addGroupByColumns(column, columns);
+    return this;
+  }
+
+  public GenericQueryBuilder select(Column column, Column... columns) {
+    genericQuery.addSelectedColumns(column, columns);
+    return this;
+  }
+
+  public String getFullSqlStatement() throws IOException {
+    return getPreparedStatement().toString();
+  }
+
+  public Records fetch() throws IOException {
+    int retryCount = 0;
+    PreparedStatement preparedStatement = getPreparedStatement();
+
+    while (true) {
+      try {
+        return getQueryResults(preparedStatement);
+      } catch (SQLRecoverableException e) {
+        if (++retryCount > MAX_CONNECTION_RETRIES) {
+          throw new IOException(e);
+        }
+      } catch (SQLException e) {
+        throw new IOException(e);
+      }
+    }
+  }
+
+  private PreparedStatement getPreparedStatement() throws IOException {
+    PreparedStatement preparedStatement = dbConnection.getPreparedStatement(genericQuery.getSqlStatement());
+    setStatementParameters(preparedStatement);
+    return preparedStatement;
+  }
+
+  private void setStatementParameters(PreparedStatement preparedStatement) throws IOException {
+    int index = 0;
+    for (WhereConstraint constraint : genericQuery.getWhereConstraints()) {
+      for (Object parameter : constraint.getParameters()) {
+        if (parameter == null) {
+          continue;
+        }
+        try {
+          preparedStatement.setObject(++index, parameter);
+        } catch (SQLException e) {
+          throw new IOException(e);
+        }
+      }
+    }
+  }
+
+  private Records getQueryResults(PreparedStatement preparedStatement) throws SQLException {
+    ResultSet queryResultSet = null;
+
+    try {
+      queryResultSet = preparedStatement.executeQuery();
+      Records results = new Records();
+      while (queryResultSet.next()) {
+        Record record = parseResultSet(queryResultSet);
+        if (record != null) {
+          results.addRecord(record);
+        }
+      }
+      return results;
+    } catch (SQLRecoverableException e) {
+      dbConnection.resetConnection();
+      throw e;
+    } finally {
+      try {
+        if (queryResultSet != null) {
+          queryResultSet.close();
+        }
+        preparedStatement.close();
+      } catch (SQLRecoverableException e) {
+        dbConnection.resetConnection();
+      } catch (SQLException e) {
+        // ignore
```

Comment:
It would be a good idea to incorporate some of the logging improvements that Tenzing added in his PRs today.